### DISCUSSION TITLE
"In-cart" quantity inaccuracies when the product has attributes

### DISCUSSION THREAD

**Discussion by lat9** [zencart/zencart#5857] [Polls]
When a product has attributes, the display of the "items-in-cart" value on a product's information page (e.g. `product_info`) can be misleading to customers, since the shopping-cart class (which provides that count) looks at its `contents` by `uprid` when making the determination as to whether or not the product is present in the cart.

Multiple `uprid` values can be present in the cart, but only the count of a specific `uprid` is returned.  If an attributed product's `product_info` page is viewed from a listing, for instance, after a quantity of some variant is added to the cart, the in-cart quantity never shows.

To reproduce, using the demo product **Matrox G400 32MB** (/index.php?main_page=product_info&cPath=1_4&products_id=2)

1. Navigate to the product's page using the base link above.  Choose the "Premium" model with "16mb" of Memory and add 1 to the cart.
2. Navigate to the `shopping_cart` page and click the product's (now) `uprid` link (/index.php?main_page=product_info&products_id=2:580a6355c0e0ed74cdb093b81ef4b530).
3. The quantity-in-cart displays as 1 (correct).
5. Navigate back to the original product link (no uprid indicated) and see that nothing shows for the items-in-cart.

**Comment by dbltoe** [comment]
Interesting for sure.  Same response in 157d.  I hate to admit I never knew the quantity in cart existed.

Would it be presumptuous of me to tag onto this that this would also be a good time to piggyback a switch or setting that allows the image in the cart to be that of the main/selected attribute versus the product's image?  I think it would be especially helpful with any product that has color as a choosable item. 

Asking for a friend 😃

**Comment by lat9** [comment]
> Interesting for sure. Same response in 157d. I hate to admit I never knew the quantity in cart existed.

Yep, it probably goes back to the time of the dinosaurs!

> Would it be presumptuous of me to tag onto this that this would also be a good time to piggyback a switch or setting that allows the image in the cart to be that of the main/selected attribute versus the product's image? I think it would be especially helpful with any product that has color as a choosable item.

It _would_; that's a totally separate issue/change-request.

**Comment by dbltoe** [comment]
No problem.  Working on and hope to have a solution soon for the image and happy to troubleshoot thoughts on this matter.

**Comment by mc12345678** [comment]
For what it's worth, this is true across the board, mixed on/off, selection of one product's attribute combination or more than one combination of that product's attributes. 

For stores that stay on the product upon add-to-cart do not update the page link to include the hashed attributes and therefore where product attribute selection is required the "quantity in cart" text is not shown.

If a product does not mandate attributes to be selected that is then added to the cart, then the quantity in cart will show for that many of product in the cart without attributes. (E.g.: /index.php?main_page=product_info&cPath=54_58&products_id=132)

The "big" question is what to do about it? Show all of the selections made? Offer ability to select to edit? hide this excess content because it takes up additional screen space? Show some semblance of the total: e.g. 2 of the 5 of this product, show a total quantity of all variants? 

Would think feedback would be appreciated on a direction, desired content to be available or possibly available and/or without touching core code a means to easily modify whatever is incorporated.

**Comment by lat9** [comment]
> The "big" question is what to do about it? Show all of the selections made? Offer ability to select to edit? hide this excess content because it takes up additional screen space? Show some semblance of the total: e.g. 2 of the 5 of this product, show a total quantity of all variants?

@mc12345678, that's exactly why I posted this as an issue as opposed to a PR'd "done-deal".  There are as many variations that could be used in the display as there are opinions as to which one is the "best".

Here are _my_ opinionated answers to your suggestion:
- _Show all of the suggestions made?_  That has the possibility of taking up inordinate amounts of 'real-estate' for complex products.
- _Offer ability to select to edit?_  The customer already has the option to select various option-combinations on a product's details page or they could go to the shopping-cart page and update the quantity for previously-selected option-combinations.
- _hide this excess content because it takes up additional screen space?_ Already an option via `Category/Products :: Product Types`' `Layout Options :: Show Quantity in Shopping Cart` configuration setting.
- _Show some semblance of the total?_ This would be my preference, identifying the **overall** quantity of the product in the cart.
